fl4health/clients/deep_mmd_clients/ditto_deep_mmd_client.py

```diff
@@ -0,0 +1,201 @@
+
+        return {"global": global_preds, "local": local_preds}, features
+
+    def _maybe_checkpoint(self, loss: float, metrics: Dict[str, Scalar], checkpoint_mode: CheckpointMode) -> None:
+        # Hooks need to be removed before checkpointing the model
+        self.local_feature_extractor.remove_hooks()
+        super()._maybe_checkpoint(loss=loss, metrics=metrics, checkpoint_mode=checkpoint_mode)
+
+    def compute_evaluation_loss(
+        self,
+        preds: Dict[str, torch.Tensor],
+        features: Dict[str, torch.Tensor],
+        target: torch.Tensor,
+    ) -> EvaluationLosses:
+        """
+        Computes evaluation loss given predictions (and potentially features) of the model and ground truth data.
+        We also need to set the training flag of the deep_mmd_losses to False to ensure that the model is in eval mode.
+
+        Args:
+            preds (Dict[str, torch.Tensor]): Prediction(s) of the model(s) indexed by name. Anything stored
+                in preds will be used to compute metrics.
+            features: (Dict[str, torch.Tensor]): Feature(s) of the model(s) indexed by name.
+            target: (torch.Tensor): Ground truth data to evaluate predictions against.
+
+        Returns:
+            EvaluationLosses: an instance of EvaluationLosses containing checkpoint loss and additional losses
+                indexed by name.
+        """
+        for layer in self.flatten_feature_extraction_layers.keys():
+            self.deep_mmd_losses[layer].training = False
+        return super().compute_evaluation_loss(preds, features, target)
+
+    def compute_loss_and_additional_losses(
+        self,
+        preds: Dict[str, torch.Tensor],
+        features: Dict[str, torch.Tensor],
+        target: torch.Tensor,
+    ) -> Tuple[torch.Tensor, Dict[str, torch.Tensor]]:
+        """
+        Computes the loss and any additional losses given predictions of the model and ground truth data.
+
+        Args:
+            preds (Dict[str, torch.Tensor]): Prediction(s) of the model(s) indexed by name.
+            features (Dict[str, torch.Tensor]): Feature(s) of the model(s) indexed by name.
+            target (torch.Tensor): Ground truth data to evaluate predictions against.
+
+        Returns:
+            Tuple[torch.Tensor, Dict[str, torch.Tensor]]; A tuple with:
+                - The tensor for the total loss
+                - A dictionary with `local_loss`, `global_loss`, `total_loss` and, based on client attributes set
+                from server config, also `deep_mmd_loss`, `feature_l2_norm_loss` keys and their respective calculated
+                values.
+        """
+        total_loss, additional_losses = super().compute_loss_and_additional_losses(preds, features, target)
+
+        if self.deep_mmd_loss_weight != 0:
+            # Compute DEEP-MMD loss
+            total_deep_mmd_loss = torch.tensor(0.0, device=self.device)
+            for layer in self.flatten_feature_extraction_layers.keys():
```

Comment:
As above, since we're not accessing components of `flatten_feature_extraction_layers`, it might be more straightforward to do:
```
for layer, layer_loss_module in self.deep_mmd_losses.items():
    layer_deep_mmd_loss = layer_loss_module(
        features[layer], features[" ".join(["init_global", layer])]
    )
    additional_losses["_".join(["deep_mmd_loss", layer])] = layer_deep_mmd_loss
    total_deep_mmd_loss += layer_deep_mmd_loss
```